linkerd/tracing/src/stream.rs

```diff
@@ -0,0 +1,365 @@
+use slab::Slab;
+use std::{
+    io,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, Weak,
+    },
+};
+use thingbuf::{
+    mpsc::{self, RecvRef, SendRef},
+    recycling::WithCapacity,
+};
+use tracing::{
+    span,
+    subscriber::{Interest, Subscriber},
+};
+use tracing_subscriber::{
+    filter::{EnvFilter, Filtered},
+    fmt::{self, format, time::SystemTime, writer::MakeWriter},
+    layer::{Context, Filter, Layer},
+    registry::LookupSpan,
+    reload,
+};
+
+/// The receiver end of a log stream.
+#[derive(Debug)]
+pub struct Reader<S: Subscriber + for<'a> LookupSpan<'a>> {
+    /// The channel reciever of log messages.
+    rx: mpsc::Receiver<Vec<u8>, WithCapacity>,
+
+    /// Shared state for this stream.
+    ///
+    /// Currently, the shared state contains a counter of how many messages have
+    /// been dropped because the channel was at capacity.
+    shared: Arc<Shared>,
+
+    /// Index in the slab of writers on the tx side of the stream.
+    ///
+    /// This is used to remove the stream when the reader is dropped.
+    idx: usize,
+
+    /// The handle for modifying the log stream layer.
+    ///
+    /// This is used to remove the stream when the reader is dropped.
+    handle: StreamHandle<S>,
+}
+
+pub(crate) type StreamLayer<S> = Filtered<WriterLayer<S>, StreamFilter, S>;
+
+/// A handle for starting new log streams, and removing old ones.
+#[derive(Debug)]
+
+pub struct StreamHandle<S> {
+    handle: reload::Handle<StreamLayer<S>, S>,
+    channel_capacity: usize,
+    channel_settings: WithCapacity,
```

Comment:
it's not really clear to me how these two capacities relate to each other. Is one of these "number of lines" to buffer and the other is "max size of a line?"

Reply:
i'll add a comment

Reply:
8fb8bc8c095022686c8dd7195393535fecb749d4 adds docs for this.